src/large_pages/node_large_page.cc

```diff
@@ -110,72 +111,63 @@ inline uintptr_t hugepage_align_down(uintptr_t addr) {
+  if (info->dlpi_name[0] == 0) {
+    for (int idx = 0; idx < info->dlpi_phnum; idx++) {
+      const ElfW(Phdr)* phdr = &info->dlpi_phdr[idx];
+      if (phdr->p_type == PT_LOAD && (phdr->p_flags & PF_X)) {
+        auto dl_params = static_cast<dl_iterate_params*>(data);
+        uintptr_t start = info->dlpi_addr + phdr->p_vaddr;
+        uintptr_t end = start + phdr->p_memsz;
+
+        if (dl_params->reference_sym >= start &&
+            dl_params->reference_sym <= end) {
+          dl_params->start = start;
+          dl_params->end = end;
+          return 1;
+        }
+      }
+    }
+  }
+  return 0;
+}
+#endif  // defined(__linux__)
 
 struct text_region FindNodeTextRegion() {
   struct text_region nregion;
   nregion.found_text_region = false;
 #if defined(__linux__)
-  std::ifstream ifs;
-  std::string map_line;
-  std::string permission;
-  std::string dev;
-  char dash;
-  uintptr_t start, end, offset, inode;
-  uintptr_t node_text_start = reinterpret_cast<uintptr_t>(&__node_text_start);
+  dl_iterate_params dl_params = {
+    0, 0, reinterpret_cast<uintptr_t>(&__node_text_start)
+  };
   uintptr_t lpstub_start = reinterpret_cast<uintptr_t>(&__start_lpstub);
 
-  ifs.open("/proc/self/maps");
-  if (!ifs) {
-    PrintWarning("could not open /proc/self/maps");
-    return nregion;
-  }
-
-  while (std::getline(ifs, map_line)) {
-    std::istringstream iss(map_line);
-    iss >> std::hex >> start;
-    iss >> dash;
-    iss >> std::hex >> end;
-    iss >> permission;
-    iss >> offset;
-    iss >> dev;
-    iss >> inode;
-
-    if (inode == 0)
-      continue;
-
-    std::string pathname;
-    iss >> pathname;
-
-    if (permission != "r-xp")
-      continue;
-
-    if (node_text_start < start || node_text_start >= end)
-      continue;
-
-    start = node_text_start;
-    if (lpstub_start > start && lpstub_start <= end)
-      end = lpstub_start;
-
-    char* from = reinterpret_cast<char*>(hugepage_align_up(start));
-    char* to = reinterpret_cast<char*>(hugepage_align_down(end));
-
-    if (from >= to)
-      break;
-
-    size_t size = to - from;
-    nregion.found_text_region = true;
-    nregion.from = from;
-    nregion.to = to;
-    nregion.total_hugepages = size / hps;
-
-    break;
+  if (dl_iterate_phdr(FindMapping, &dl_params) == 1) {
+    dl_params.start = dl_params.reference_sym;
+    if (lpstub_start > dl_params.start && lpstub_start <= dl_params.end)
+      dl_params.end = lpstub_start;
+
+    if (dl_params.start < dl_params.end) {
+      char* from = reinterpret_cast<char*>(hugepage_align_up(dl_params.start));
+      char* to = reinterpret_cast<char*>(hugepage_align_down(dl_params.end));
+      if (from < to) {
+        size_t pagecount = (to - from) / hps;
+        if (pagecount > 0) {
+          nregion.found_text_region = true;
+          nregion.from = from;
+          nregion.to = to;
+          nregion.total_hugepages = pagecount;
+        }
+      }
+    }
```

Comment:
@bnoordhuis I added some more sanity checks to ensure that we create a region that has proper offsets and > 0 page count.

Reply:
... or rather, that we consider the region to have been found only if it has proper offsets and >0 page count.